### PR TITLE
119 surgical tool segmentation

### LINKED ISSUE #119: Surgical Tool Segmentation Bundle
**Is your feature request related to a problem? Please describe.**
Surgical tool segmentation is very useful in endscope videos. It can highlight the surigical tools from body tissue and help doctors to localize tools. 

**Describe the solution you'd like**
A sugical tool segmentation bundle and models to implement:

1. Tool segmentation training
2. Tool segmentation inference
3. Tool Segmentation evaluate

### PR DESCRIPTION
**Pull request by binliunls**
Fixes #119  .

### Description
A surigical tool segmentation bundle which contains pretrained models. Training, inference and evaluate can be done using this bundle given dataset.

### Status
**Work in progress**

### Please ensure all the checkboxes:

- [ ] Codeformat tests passed locally by running `./runtests.sh --codeformat`.
- [ ] In-line docstrings updated.
- [ ] Update `version` and `changelog` in `metadata.json`.
- [ ] Ensure versions of packages such as `monai`, `pytorch` and `numpy` are correct in `metadata.json`.
- [ ] Descriptions should be consistent with the content, such as `eval_metrics` of the provided weights and TorchScript modules.
- [ ] Files larger than 25MB are excluded and replaced by providing download links in `large_file.yml`.
- [ ] Avoid using path that contains personal information within config files (such as use `/home/your_name/` for `"bundle_root"`).

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @binliunls ,

Please reopen this PR and update for review.

Thanks.

File: models/surgical_tool_segmentation/configs/train.json [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+        },
+        "inferer": {
+            "_target_": "SimpleInferer"
+        },
+        "handlers": [
+            {
+                "_target_": "ValidationHandler",
+                "validator": "@validate#evaluator",
+                "epoch_level": true,
+                "interval": 1
```
Comment (yiheng-wang-nv) [inline models/surgical_tool_segmentation/configs/train.json]: Hi @binliunls , please define a variable called `val_interval`, see also: https://github.com/Project-MONAI/model-zoo/blob/1912aba5a6211c9b49510735180fbecf6489131c/CONTRIBUTING.md#preferred-files-and-keys
It is a requirement from monai FL side.

Comment (binliunls) [reply models/surgical_tool_segmentation/configs/train.json]: Thanks @yiheng-wang-nv , I have added the val_interval parameter to the train.json

File: models/surgical_tool_segmentation/configs/metadata.json [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+    },
+    "monai_version": "1.0.0",
+    "pytorch_version": "1.12.0",
+    "numpy_version": "1.22.4",
+    "optional_packages_version": {
+        "nibabel": "4.0.1",
+        "pytorch-ignite": "0.4.9"
+    },
+    "task": "surgical tool segmentation",
+    "description": "A pre-trained binary segmentation model for surgical tool segmentation",
```
Comment (yiheng-wang-nv) [inline models/surgical_tool_segmentation/configs/metadata.json]: Thanks @binliunls , this bundle looks good to me.
Hi @Nic-Ma and @wyli , since the dataset is private, do you think it is better to add some figures to visualize what kinds of tools can be recognized based on this model? For example, do inference on some similar and public datasets (such as [this bundle](https://github.com/Project-MONAI/model-zoo/tree/dev/models/ventricular_short_axis_3label) does).

**Review by yiheng-wang-nv — APPROVED** [review]

File: models/surgical_tool_segmentation/configs/metadata.json (hunk 2 above)
Comment (Nic-Ma) [reply models/surgical_tool_segmentation/configs/metadata.json]: Interesting idea!
@binliunls Could you please help add inference example with the public dataset in the video tutorial?

Thanks in advance.

Comment (binliunls) [reply models/surgical_tool_segmentation/configs/metadata.json]: Hi @Nic-Ma , @yiheng-wang-nv 
     I have added the EndoVis 2017 dataset to the readme file.

**Review by Nic-Ma — APPROVED** [review]

**Review by Nic-Ma — CHANGES_REQUESTED** [review]
Pendng for legal confirmation for the lisence.

Thanks.

**Comment by tangy5** [comment]
hi @binliunls , This is exciting, I'm curious whether this bundle is mostly ready. 
I'm working on MONAI Label endoscopy app to support loading models directly from bundle instead of python codes. 
I will test the development with this bundle if it's working. No need to rush for making it perfect, just let me know if this is usable in a development stage. 
Thank you!

**Comment by binliunls** [comment]
> hi @binliunls , This is exciting, I'm curious whether this bundle is mostly ready. I'm working on MONAI Label endoscopy app to support loading models directly from bundle instead of python codes. I will test the development with this bundle if it's working. No need to rush for making it perfect, just let me know if this is usable in a development stage. Thank you!

Hi @tangy5 ,
       Most parts of this bundle are done. We are just waiting for the legal confirmation.
Thanks,
BIn

File: models/surgical_tool_segmentation/docs/readme.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
+# Description
+A pre-trained model for the surgical tool segmentation task.
+
+# Model Overview
+This model is trained using a flexible unet structure with an efficient-b0 [1] as the backbone and a UNet architecture [2] as the decoder. Datasets use private samples from Activ-Surgical.
```
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/docs/readme.md]: Let's have a consistent reference to `Activ Surgical` and remove the `-`. Perhaps also a link to their [website](https://www.activsurgical.com/)?

Comment (binliunls) [reply models/surgical_tool_segmentation/docs/readme.md]: Hi @jin-nvidia ,
     Thank you for your advice. I have updated the reference to Activ Surgical and added their webseite link to it.
Thanks,
Bin

File: models/surgical_tool_segmentation/docs/readme.md [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+
+# Model Overview
+This model is trained using a flexible unet structure with an efficient-b0 [1] as the backbone and a UNet architecture [2] as the decoder. Datasets use private samples from [Activ Surgical](https://www.activsurgical.com/).
+The [pytorch model](https://drive.google.com/file/d/14r6WmzaZrgaWLGu0O9vSAzdeIGVFQ3cs/view?usp=sharing) and [torchscript model](https://drive.google.com/file/d/1i-e5xXHtmvmqitwUP8Q3JqvnmN3mlrEm/view?usp=sharing) are shared in google drive. Details can be found in large_files.yml file.
+
+## Data
+Datasets used in this work were provided by [Activ Surgical](https://www.activsurgical.com/).
+
+Since datasets are private, existing public datasets like [EndoVis 2017](https://endovissub2017-roboticinstrumentsegmentation.grand-challenge.org/Data/) can be used to train a similar model.
+
```
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/docs/readme.md]: Can we add a reference to what needs to be changed in configs/train.json when users bring their own data? The EndoVis 2017 are not of jpg format but png. For example, `"dataset_dir", "images", "labels",  "val_images",  "val_labels"` all need to change, anything else?

File: models/surgical_tool_segmentation/docs/readme.md [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+## Training configuration
+The training was performed with an at least 12GB-memory GPU.
+
+Actual Model Input: 736 x 480 x 3
+
+## Input and output formats
```
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/docs/readme.md]: Suggest to caution users when their raw images aren't in the jpg format or when their image/mask pairs aren't of the same format, they may need to change the `"train":"deterministic_transforms"` in configs/train.json due to different dataloaders for different image types. For example if my data has png images and npy masks, the same transforms will not work.

File: models/surgical_tool_segmentation/docs/readme.md [hunk 6]
```diff
@@ -0,0 +1,95 @@
+# Description
+A pre-trained model for the surgical tool segmentation task.
+
+# Model Overview
+This model is trained using a flexible unet structure with an efficient-b0 [1] as the backbone and a UNet architecture [2] as the decoder. Datasets use private samples from [Activ Surgical](https://www.activsurgical.com/).
+The [pytorch model](https://drive.google.com/file/d/14r6WmzaZrgaWLGu0O9vSAzdeIGVFQ3cs/view?usp=sharing) and [torchscript model](https://drive.google.com/file/d/1i-e5xXHtmvmqitwUP8Q3JqvnmN3mlrEm/view?usp=sharing) are shared in google drive. Details can be found in large_files.yml file.
```
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/docs/readme.md]: Suggest to add note: 

Modify the path specified in `configs/train.json` and `configs/inference.json`: `"bundle_root"` to reflect where the models are downloaded. Under the path specified by `"bundle_root"` the expected directory structure is `models/model.pt`.

File: models/surgical_tool_segmentation/docs/readme.md (hunk 4 above)
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/docs/readme.md]: Can we add a reference to what needs to be changed in configs/inference.json when users bring their own data? For example, `"dataset_dir", "datalist"` need to change, anything else?

File: models/surgical_tool_segmentation/docs/readme.md [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+
+Export TensorRT float16 model from the onnx model:
+
+```
+trtexec --onnx=models/model.onnx --saveEngine=models/model.trt --fp16 --minShapes=INPUT__0:1x3x736x480 --optShapes=INPUT__0:4x3x736x480 --maxShapes=INPUT__0:8x3x736x480 --shapes=INPUT__0:4x3x736x480
+```
+This command need TensorRT with correct CUDA installed in the environment. For the detail of installing TensorRT, please refer to [this link](https://docs.nvidia.com/deeplearning/tensorrt/install-guide/index.html). In addition, there are padding operations in this FlexibleUNet structure that not support by TensorRT. Therefore, when tried to convert the onnx model to a TensorRT engine, an extra step shown below is needed to execute. The "new_model.onnx" should be the "model.onnx" used to do converting.
```
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/docs/readme.md]: Suggest to move this polygraphy step before the previous trtexec step and change the onnx model name in the trtexec command.

File: models/surgical_tool_segmentation/docs/readme.md [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+Export checkpoint to onnx file, which has been tested on pytorch 1.12.0:
+
+```
+python scripts/export_to_onnx.py --model models/model.pt --outpath models/model.onnx
+```
+
+Export TorchScript file to a torchscript module targeting a TensorRT engine with float16 precision.
+
+```
+torchtrtc -p f16 models/model.ts models/model_trt.ts "[(1,3,736,480);(4,3,736,480);(8,3,736,480)]"
```
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/docs/readme.md]: Following the advice of https://github.com/pytorch/TensorRT#nvidia-ngc-container, and using the container `nvidia/pytorch:22.09-py3`, still encountered error 
> torchtrtc -p f16 models/model.ts models/model_trt.ts "[(1,3,736,480);(4,3,736,480);(8,3,736,480)]"
> terminate called after throwing an instance of 'c10::Error' 
>  what():  Trying to create tensor with negative dimension -1: [-1, 3, 736, 480]

Are there any additional steps within the container needed?

File: models/surgical_tool_segmentation/configs/train.json [hunk 9]
```diff
@@ -0,0 +1,275 @@
...
+                        2
+                    ],
+                    "keys": [
+                        "pred",
+                        "label"
+                    ]
+                }
+            ]
+        },
+        "trainer": {
```
Comment (jin-nvidia) [inline models/surgical_tool_segmentation/configs/train.json]: In this current setting is it possible to add an option to load a pre-trained checkpoint?

File: models/surgical_tool_segmentation/docs/readme.md (hunk 4 above)
Comment (binliunls) [reply models/surgical_tool_segmentation/docs/readme.md]: Hi @jin-nvidia ,
     Thank you for your advice. I have added a description of how to use user owned dataset in this bundle.
Thanks,
Bin

File: models/surgical_tool_segmentation/docs/readme.md (hunk 5 above)
Comment (binliunls) [reply models/surgical_tool_segmentation/docs/readme.md]: Hi @jin-nvidia ,
     Thank you for your advice. I have added a notice of data format in this bundle readme file.
Thanks,
Bin

File: models/surgical_tool_segmentation/docs/readme.md (hunk 6 above)
Comment (binliunls) [reply models/surgical_tool_segmentation/docs/readme.md]: Hi @jin-nvidia ,
     Thank you for your advice. This note has been added to Model Overview part.
Thanks,
Bin

File: models/surgical_tool_segmentation/docs/readme.md (hunk 7 above)
Comment (binliunls) [reply models/surgical_tool_segmentation/docs/readme.md]: Hi @jin-nvidia ,
     Thank you for your advice. Polygraphy step was moved.
Thanks,
Bin

File: models/surgical_tool_segmentation/docs/readme.md (hunk 8 above)
Comment (binliunls) [reply models/surgical_tool_segmentation/docs/readme.md]: Hi @jin-nvidia ,
     This error may caused by hardware and software problems. You can add "-v" option first to check the detail.
Thanks,
Bin

File: models/surgical_tool_segmentation/docs/readme.md (hunk 4 above)
Comment (binliunls) [reply models/surgical_tool_segmentation/docs/readme.md]: Hi @jin-nvidia ,
     Thank you for your advice. A description of what need to do with user dataset was added to the readme file.
Thanks,
Bin

**Review by Nic-Ma — APPROVED** [review]
